### PR TITLE
Fix a memory leak when running the unit tests.

### PR DESCRIPTION
**Pull request by reivilibre**
Global state is evil, but it's a bit much right now to rework this.

Introduced in #13717...

> Running unit tests involves starting loads of fake homeservers back-to-back.
> Each time the fake homeserver is started, we register the caches into the global cache metrics collector/state.
> My change makes it so that we append a function to a list, rather than setting a key in a dictionary. This means that we don't clean up old entries anymore, because the list is never emptied, whereas setting a key in a list overwrites what's already there, so these entries got 'cleaned up' by the new entries replacing them.

This fix is quite important as otherwise running the test suite can deplete your system of its memory :-).

### CODE REVIEW

File: synapse/util/caches/__init__.py [hunk 1]
```diff
@@ -205,8 +205,9 @@ def register_cache(
         add_resizable_cache(cache_name, resize_callback)
 
     metric = CacheMetric(cache, cache_type, cache_name, collect_callback)
+    metric_name = "cache_%s_%s" % (cache_type, cache_name)
```
Comment (reivilibre) [inline synapse/util/caches/__init__.py:208]: this is reintroduced from before #13717

File: synapse/util/metrics.py [hunk 2]
```diff
@@ -220,21 +220,21 @@ class DynamicCollectorRegistry(CollectorRegistry):
...
         yield from super().collect()
 
-    def register_hook(self, hook: Callable[[], None]) -> None:
+    def register_hook(self, metric_name: str, hook: Callable[[], None]) -> None:
         """
         Registers a hook that is called before metric collection.
         """
 
-        self._pre_update_hooks.append(hook)
+        self._pre_update_hooks[metric_name] = hook
```
Comment (clokep) [inline synapse/util/metrics.py:240]: Should we be doing anything if we're overriding one? Do we need to "unregister" the old one somehow?

Overall this seems to rollback to the previous behavior though so is likely OK.

**Review by clokep — APPROVED** [review]

File: synapse/util/metrics.py (hunk 2 above)
Comment (reivilibre) [reply synapse/util/metrics.py:240]: It will be 'unregistered' by being overwritten in the dict; the hook isn't tracked anywhere else.

Comment (clokep) [reply synapse/util/metrics.py:240]: Ah-ha! I thought it somehow got tracked internally from the tracing bits, but doesn't look like it. 👍 Thank you for the explanation!

**Pull request merged**